### ISSUE TITLE
[Enphase] Discrepancy issue between two production JSON's

### ISSUE THREAD

**Issue by fwieringen** [domoticz/domoticz#5734]
Hi,

I noticed a discrepancy in the Enphase Envoy. It appears to be a bug in the Envoy itself. My total lifetime production is around 4.1MWh. However, the Envoy Installer screen shows 588 kWh. The Enphase website shows 4.1MWh.

The lifetime production can be requested using two API's:

- api/v1/production/ (this one shows the 4.1 MWh)
- production.json?details=1 (this one shows the faulty 588 kWh)

I noticed that the latest Enphase updates in Domoticz use the second method to get the total lifetime production. 

I'm not sure if I'm the only one having this issue, but perhaps it's an idea to use the api/v1/production method to get the total production, because that shows the correct one.

I'm having a chat with Enphase tech support as well, so will update this thread accordingly.

Best regards, Friso

**Comment by gizmocuz** [comment]
Also mention please if you are using the latest beta version of Domoticz.
As since a few versions we use the second data from production.json?details=1

**Comment by fwieringen** [comment]
Yes, using Domoticz build 15403. I noted this one uses the production.json?details=1 variant and this is the one showing the wrong counter.

Enphase is also looking into the discrepancy. Apparently the production.json?details=1 is the json used to feed the installer screen with data and the api/v1/production method is used to show daily, weekly and lifetime production.

**Comment by gizmocuz** [comment]
Is all sorted out now? Do you get the right values?

**Comment by fwieringen** [comment]
Hi,

Haven't heard back from Enphase yet. Updated to the latest version of Domoticz, but I assume that one uses the production.json?details=1 url right?

In that case, the situation is still the same.

![image](https://github.com/domoticz/domoticz/assets/4232879/37505d18-1bee-4b52-a715-f7f318055acf)

Best regards,
Friso

**Comment by gizmocuz** [comment]
Yes, we use 'production.json?details=1'
And this one is used for quite a long time and also with all other libraries.
But depending on if you have a metered gateway we will use the first or second result from this query (second in case of metered)
Best to look in our code base function -> **EnphaseAPI::getProductionDetails** and **EnphaseAPI::parseProduction**

But this value cannot be negative! This is always positive.

It seems (the) envoy is sometimes resetting it's counter value to 0, maybe by a reboot/firmware upgrade

In your situation, shift left click on the data point in the month chart, you will get a popup, and select 'delete'
And your negative value should be gone...
If it is a metered envoy with CT Clamps, a little bit negative usage can be displayed (like -2 or -3 watt), but the counter value should be ok

**Comment by fwieringen** [comment]
Hi,

Thanks. Yes, I think it's really a bug/discrepancy in the Envoy itself. The results of both JSON's are not aligned.

![image](https://github.com/domoticz/domoticz/assets/4232879/747e8061-abe0-4424-81b0-ee5be76ef28d)

The reason why I thought it maybe a change in Domoticz as well is because it worked quite ok in Domoticz up until July 6th... Even though the Envoy already showed this lower lifetime usage in the Installer screen since April.

Anyway, I'll wait on the response of Enphase and use my own NodeRed flow in the meantime to pull the lifetime wH from the other JSON.

Thanks!

**Comment by reinos** [comment]
Hi @fwieringen please let me know when you receive any word/update from Enphase as some of my friends do have the same issue (or sort of) with Enphase. 

It seems that occasionally the `wattHoursLifetime` is reset to `0`. The website (cloud) of Enphase is showing the correct values, but all internal values (`https://enphase.local/home#overview` `https://enphase.local/api/v1/production` `https://enphase.local/production.json`)  are reset to `0`.  

I personally also have Enphase (IQ8) and it runs without any issue. The main difference with the other systems of my friends is the type of inverter. They have the IQ7.

What we also discover is that IQ7 is running on an older version than the IQ8 inverters. However, it is not possible to upgrade the firmware to the same version that are running on the IQ8 inverters. 
So basically you are stuck with an older firmware on the IQ7 that seems to have a bug that reset the `wattHoursLifetime` occasionally.

Perhaps you can use this info in your chat with Enphase.

**Comment by fwieringen** [comment]
I received a preliminary answer from Enphase. They basically confirm that there's a discrepancy between two counters in the Envoy, but they cannot explain why. 

The /api/v1/production value shows the correct value and is in line with the value on the Enphase website. The /production.json has been reset to 0 on May 28th in my case. 

I asked a follow up question what causes this. They're still working on that.

In my case I created a workaround with a Node Red flow to get the production value (and inverter values) from the right API and feed it via MQTT to Domoticz. But I am thinking of enhancing that Node Red flow to be able to add a hard coded correction in case the other counter will also reset to 0 ;-)

**Comment by tuinstoelen** [comment]
Monitoring this thread. I've got the same issue - the whLifetime counter in production.json resets to zero. Happened somewhere today.

IQ7a inverters with firmware version 520-00082-r01-v04.27.04 / 540-00169-r01-v04.27.09 with an Envoy-S-Standard-EU, firmware version D7-6-175. 

@fwieringen if you get an update or a fix from Enphase I'd very much like to hear it!

**Comment by gizmocuz** [comment]
I experienced this already twice... But then I was also updating the firmware to a new version, and I had to turn off the grid a few times. Maybe it's related

**Comment by fwieringen** [comment]
Heard back from Enphase, although not very satisfying. Their story is that the production.json is used in the installer screen and the installer isn't interested in the lifetime production. He just wants to see if the site is currently producing. They call it 'today's lifetime production'. 

They didn't have an explanation on why the api/v1/production still contains the value which has not been reset. Apparently there's another counter in there that stores this.

**Comment by tuinstoelen** [comment]
Perhaps we can switch over to using the wattsNow and wattHoursLifetime from the v1 api, and not use the production.json (at least for these two data points).

Just slapping in another web request to fetch the wattsNow and wattHoursLifetime from the v1 api might not be the prettiest approach, as it seems more data is used from the production details json (in case of a metered version?).

I'm not sure what the metered version is, i do not think i have it. This makes me a little unsure about the necessary changes to make it a 'pretty' fix, instead of just slapping on fetch to the v1 api. 

Any guidance is appreciated.

**Comment by gizmocuz** [comment]
The 'metered' version is a envoy where CT clamps have been installed

**Comment by reinos** [comment]
The only device that facing this issue is the `IQ7`. The `IQ8` with firmware has no issues at all. 

Also, for some of installation `https://envoy.local/api/v1/production` seems stable, but for not for all. I faced this issue at the installation of a friend. For one installation this endpoint is stable, but for another installation with the same firmware version `D7.0.88` it is not working, values are reset to 0 once in a while. 

Upgrading the IQ7 (current version D7.0.88) to the same version that runs on my IQ8 (D7.6.168) is not possible. The installer is not offer a newer firmware option.

My simple conclusion, this should be fixed by Enphase and not within Domoticz. :-(

**Comment by fwieringen** [comment]
Or anticipate in Domoticz for this behaviour/bug and automatically apply a correction when the new received value for lifetime production is less than the previously received value ;-)

E.g. when the previous production is 1000 and the new value = 1, add a correction of 1000 and set the counter to 1001.

**Comment by tuinstoelen** [comment]
Or always fetch both, and use the larger of the two. Could that work?

Does anyone know if there are there any potential throttling issues with the addition of another periodical request?

**Comment by gizmocuz** [comment]
I have IQ8-M and have seen this issue
But not very often, when this happens (the reset to 0), the day value is incorrect. Deleting that value the next day (shift-click on the day value to delete it), fixes it in the graph
But I agree, this should not happen of course.

@fwieringen ,

I am not a fan of fixing this like that... I have done so on multiple places in Domoticz, but it should be solved at Enphase.
It's a bug, and saying it was not meant to be counting for a lifetime is crap as the variable has the name 'lifetime'

Happy to use another API call

@reinos, I had D7.0.88 and managed to update to the latest firmware. I had to use the installer tool (not in AP mode, it downloaded the firmware), then I switched to AP mode, and it installed the firmware...
(Or sort of... I know I was struggling with this as well, but it is possible.

**Comment by gizmocuz** [comment]
> Or always fetch both, and use the larger of the two. Could that work?
> 
> Does anyone know if there are there any potential throttling issues with the addition of another periodical request?

No that's buggy code and results in multiply queries... We want to minimize this as also querying the envoy a lot can result in stale envoy and you need to restart it.

**Comment by fwieringen** [comment]
For completeness the answer of Enphase:

_
Good day! Thank you for the patience!
Please check the below update, that we got it from the software team.


"The values are different as UI page lifetime production means here is today’s total production ( today lifetime)

UI page data is taken from /ivp/pdm/production

and enlighten lifetime means whole life after the system installed and started production (since start day of production)

enlighten data is taken from /ivp/pdm/energy
Lifetime Energy shown in Enlighten seems to be correct, and it is matching with the sum of Lifetime Energy for all the microinverters, which is 4.3 MWh. It is also consistent with the installed capacity and time of operation, average of, 5750Wh for 750 days = 4.3 MWh.

Image_2023-07-28_17-12-18.png
From energy report, sum of energy from 5/28/2023 to 7/11/2023 is 575 kWh. Yesterday when we checked the lifetime energy was matching with the sum of energy from 5/28/2023 till yesterday. The Lifetime energy of the installer page is coming from envoy, which seems to have reset somewhere in the last week of May 2023. That is why it's showing less value. We also don’t see any system events in 2023 till May.”_

**Comment by gizmocuz** [comment]
Could you also post the json result of

/ivp/pdm/production
/ivp/pdm/energy

maybe another one

/ivp/peb/devstatus

Not sure if you need installer rights for these ivp url's or that it is also working for non-installer tokens

**Comment by fwieringen** [comment]
I get a 401, so probably installer rights are needed.

**Comment by gizmocuz** [comment]
That would mean we are not able to use these URL's

**Comment by tuinstoelen** [comment]
Here are two enphase support forum links about the issue:

https://support.enphase.com/s/question/0D53m00009INBskCAH/local-api-return-0-for-wh-counters-for-production-after-new-firmware-update

https://support.enphase.com/s/question/0D53m00009JjkL4CAJ/local-apis-on-d76175-f79c8d-are-not-reporting-total-watt-hours-httpsenvoylocalapiv1production-httpsenvoylocalproductionjsondetails1all-values-are-zero-why-push-a-broken-firmware-was-this-even-tested

**Comment by 0crap** [comment]
Yep, same issue here as well. (See #5782)

Few weeks ago my Envoy-S-Standard-EU was updated from `D7.0.88` to `D7.6.175`.
IQ7+ is still running on 520-v04.27.04 / 540-v04.27.10

The big negative value was show in Domoticz on a random day the Envoy-S was already running the D7.6.175 version.
No updates going on or me "playing" with the system in any way,

**Comment by gizmocuz** [comment]
@0crap , okey, but that was one time? How is it working now?

**Comment by 0crap** [comment]
Correct, happened one time.
The next day the Today counter started at zero and working fine for now.
Will keep an eye on it.

**Comment by waltervl** [comment]
No feedback so seems to work fine now.
